bridge-fungible-token/src/bridge_fungible_token.sw

```diff
@@ -0,0 +1,150 @@
+contract;
+
+dep data;
+dep errors;
+dep events;
+dep utils;
+
+use bridge_fungible_token_abi::BridgeFungibleToken;
+use contract_message_receiver::MessageReceiver;
+use core::num::*;
+use errors::BridgeFungibleTokenError;
+use events::{RefundRegisteredEvent, WithdrawalEvent};
+use std::{
+    chain::auth::{
+        AuthError,
+        msg_sender,
+    },
+    constants::ZERO_B256,
+    context::{
+        call_frames::{
+            contract_id,
+            msg_asset_id,
+        },
+        msg_amount,
+    },
+    logging::log,
+    storage::StorageMap,
+    vm::evm::evm_address::EvmAddress,
+};
+use utils::{
+    burn_tokens,
+    correct_input_type,
+    decompose,
+    input_message_data,
+    input_message_data_length,
+    input_message_recipient,
+    input_message_sender,
+    mint_and_transfer_tokens,
+    parse_message_data,
+    safe_b256_to_u64,
+    send_message_output,
+    transfer_tokens,
+    u64_to_b256,
+};
+
+////////////////////////////////////////
+// Storage declarations
+////////////////////////////////////////
+storage {
+    refund_amounts: StorageMap<(EvmAddress, EvmAddress), Option<b256>> = StorageMap {},
+}
+
+////////////////////////////////////////
+// Storage-dependant private functions
+////////////////////////////////////////
+#[storage(write)]
+pub fn register_refund(from: EvmAddress, asset: EvmAddress, amount: b256) {
+    storage.refund_amounts.insert((from, asset), Option::Some(amount));
+    log(RefundRegisteredEvent {
+        from,
+        asset,
+        amount,
+    });
+}
+
+////////////////////////////////////////
+// ABI Implementations
+////////////////////////////////////////
+// Implement the process_message function required to be a message receiver
+impl MessageReceiver for Contract {
+    #[storage(read, write)]
+    fn process_message(msg_idx: u8) {
+        let input_sender = input_message_sender(1);
+
+        require(input_sender.value == LAYER_1_ERC20_GATEWAY, BridgeFungibleTokenError::UnauthorizedSender);
+
+        let message_data = parse_message_data(msg_idx);
+
+        // @review this.
+        // Register a refund if tokens don't match ?
+        // register_refund(message_data.from, message_data.l1_asset, message_data.amount);
```

Comment:
Yes, there need to be a refund issued so a user can get their incorrect tokens back. Should hopefully never happen, but we should still cover this scenario.